### ISSUE TITLE
Respect direction tag for traffic lights and stop signs

### ISSUE THREAD

**Issue by zzptichka** [abrensch/brouter#183]
When `direction` tag is set on nodes such as `highway=stop`, `highway=give_way` and `highway=traffic_signals` BRouter should only return these tags if the direction matches the direction of travelling. 
Right now we are getting many phantom stop/yield signs in the route.
This will also prompt mappers to assign direction tags that are pretty important. 
More information about direction tag: https://wiki.openstreetmap.org/wiki/Tag:highway%3Dstop
Taginfo usage: https://taginfo.openstreetmap.org/tags/highway=stop#combinations

**Comment by Phyks** [comment]
Hi,

I could add `direction` tag in `lookups.dat` (https://github.com/abrensch/brouter/blob/master/misc/profiles2/lookups.dat) as part of #176. 

> Right now we are getting many phantom stop/yield signs in the route.

I'm not aware of any particular use case of traffic signals, give way etc in the default profiles. Would you have an example so that I update everything accordingly?

Thanks!

**Comment by abrensch** [comment]
There's a similar request on the google-group:

  https://groups.google.com/forum/#!topic/osm-android-bikerouting/EZUrFUcekHs

yes, just having the direction-tag in the lookups should do it, then you can combine that with the way-direction:

  assign isValid = xor way:reversedirection direction=forward

**Comment by zzptichka** [comment]
Adding direction tag to lookups.dat should do it.
Thanks.
Traffic lights add time to travel, especially in densely populated cities, so it would be nice to correctly count them in profiles and maybe show them on the map on frontend.
Same thing for stop signs. On my commute when I have similar options I would prefer to choose a route that doesn't pass through many stop signs.

**Comment by Phyks** [comment]
Should be done in https://github.com/abrensch/brouter/pull/176/commits/79eba309d51a4ec4338133c2511293fb098a0752 as part of #176.

**Comment by polyscias** [comment]
Good that the direction tag has been added but for traffic lights I think also the traffic_signals:direction tag should be added.

See https://wiki.openstreetmap.org/wiki/Key:traffic_signals:direction, for traffic lights that are mapped separate from the road (which can not be used by router AFAIK) the direction tag is used, but for traffic lights that are mapped on/in the road traffic_signals:direction is used, see for example 

https://www.openstreetmap.org/node/1959101729

Looking at taginfo I see that currently direction=forward/backward/both is used 547083 times and traffic_signals:direction=forward/backward/both is used 103915 times.

For brouter direction and traffic_signals:direction are equivalent I think so an alias would also be fine.

**Comment by Phyks** [comment]
@polyscias I should have added it in https://github.com/abrensch/brouter/pull/185.

I added the `traffic_signals:direction` tag directly, not as an alias of `direction` tag because, as far as I know about BRouter internals, aliases can only be easily defined on values of a same tag, not on different tags.

**Comment by abrensch** [comment]
current data is based on new lookup.dat, and Online Router is using it

**Comment by polyscias** [comment]
Thanks for updating the lookups.dat file!

Now looking on how to use it, I saw the suggestion "assign isValid = xor way:reversedirection direction=forward"

What I do not get yet is how that would work, highway=traffic_signals + traffic_signals:direction are in node context but the exact meaning of forward and backward depends on the direction the way of which the node is part is in OSM has.

I tried "way:reversedirection" in brouter-web thinking that might be the way to get this context in but I then get a "Profile error: ParseException at line 305: unknown variable: reversedirection"

**Comment by polyscias** [comment]
Found today some time to give the direction tag a try.

> assign isValid = xor way:reversedirection direction=forward

That did not work out-of-the-box:

ParseException: unknown variable: reversedirection

What works for me is setting another variable in way context and using that.

```
---context:way   # following code refers to way-tags
assign iswayreverse = reversedirection=yes

---context:node  # following code refers to node tags
assign issignalforward = xor way:iswayreverse traffic_signals:direction=forward|both
assign issignforward = xor way:iswayreverse direction=forward|both

assign initialcost
 add switch and highway=traffic_signals issignalforward 150 # Kosten für Ampel
  switch and issignforward highway=stop 40 # Kosten für Stoppschild
  switch railway=crossing|level_crossing 30 # Kosten für Bahnübergang
  0
 switch bikeaccess
  0 
 switch footaccess
  100 # Kosten fürs Absteigen
 1000000 #Kosten für verbotene oder private Wege

```

**Comment by polyscias** [comment]
The code above did not work correct yet for "legacy traffic lights", i.e. highway=traffic_lights on a crossing node with ways that come in which have no oneway tag.

Testing the code you will see that often a detour is taken, to prevent the traffic light costs, to prevent this, I updated the costs for a road with "bicycle=no|use_sidepath" to 10 and take the traffic light cost for every pedestrian traffic light taken respectively of the direction, see the changed code below.

With this I still see sometimes detours but that is either because the OSM data is not complete (cyclepath has traffic lights but parallel footway not etc.) or that it is a viable option to consider.

```
---context:way   # following code refers to way-tags

#
# calculate logical bike access
#
assign bikeaccess =
       if any_cycleroute then true
       else if bicycle= then
       (
         if bicycle_road=yes then true
         else if cycleway=yes|track|share_busway|shared_lane|lane then true # includes cyclestreet=yes (alias)
         else if vehicle= then defaultaccess
         else not vehicle=private|no
       )
       else not bicycle=private|no|dismount|use_sidepath

#
# calculate logical foot access
#
assign isroad = 
       highway=trunk|primary|secondary|tertiary|trunk_link|primary_link|secondary_link|tertiary_link|residential|unclassified

assign footaccess =
       if bikeaccess then true
       else if bicycle=dismount then true
       else if and isroad bicycle=no|use_sidepath then false # implicite foot=no
       else if foot= then defaultaccess
       else not foot=private|no

#
# if not bike-, but foot-access, just a moderate penalty,
# otherwise access is forbidden
#
assign accesspenalty =
       if bikeaccess then 0
       else if footaccess then 4
       else if and isroad bicycle=no|use_sidepath then 10
       else 10000

assign isoneway =
       if oneway:bicycle=yes then true
       else if oneway= then junction=roundabout
       else oneway=yes|true|1

assign iswayreverse = reversedirection=yes

---context:node  # following code refers to node tags

# cost for traffic lights, stop signs and railway crossings
assign trafficlightcost = 150
assign stopcost = 40
assign railwaycost = 20

assign notcycling = greater way:onewaypenalty 0 # if not cycling take all traffic lights

assign stopcost =
       # Traffic signals
       if highway=traffic_signals then
       (
         if notcycling then trafficlightcost
         # no directions set, "old style" if coming from a two-way road
         else if traffic_signals:direction= then
         (
           if way:isoneway then
           (
             switch way:iswayreverse 0 trafficlightcost
           )
           else trafficlightcost
         )
         # both
         else if traffic_signals:direction=both then trafficlightcost
         # depending on the direction we came from
         else switch xor way:iswayreverse traffic_signals:direction=forward trafficlightcost 0
       )
       # Stop sign
       else if highway=stop then
       (
         if direction= then stopcost
         else switch xor way:iswayreverse direction=forward stopcost 0
       )
       # Railway
       else if railway=crossing|level_crossing then railwaycost
       # Default
       else 0

assign initialcost =
       if bikeaccess then stopcost
       else if footaccess then add 100 stopcost
       else 1000000
```

**Comment by Phyks** [comment]
@polyscias Nice! Did you manage to fix the issues we were also discussing in #79 about traffic light penalties?

If this is working fine without any major artifacts, it's probably worth wrapping this in a PR on the bike profile. It will be easier to test and probably worth a merge if there are not much artifacts (although maybe having it disabled by default with a global switch to enable this avoid traffic light behavior)

**Comment by polyscias** [comment]
Without artifacts? It depends on what you call this situation:

![Screenshot_20191022_141943](https://user-images.githubusercontent.com/6134878/67285022-4f246d00-f4c6-11e9-814f-2dd061970481.png)

The result for [this link](http://brouter.de/brouter-web/#map=19/52.08529/4.87423/standard&lonlats=4.874722,52.085441;4.873729,52.085159).

What brouter does here makes sense, the traffic light "adds length" so the detour is shorter.

If I see this situation I typically open up JOSM see if the detour makes sense and if not, add a "only_straight_on" restriction:

![Screenshot_20191022_143022](https://user-images.githubusercontent.com/6134878/67285872-e5a55e00-f4c7-11e9-9288-13b59ad29fac.png)

Downloading then the updated .rd5 file next day I see brouter taking the normal route.

So this is the case I was talking about where I think the OSM data is not complete (like the  cyclepath has traffic lights but parallel footway)

> Did you manage to fix the issues we were also discussing in #79 about traffic light penalties?

Yes, that was easy, just setting the accesspenatily using "if and isroad bicycle=no|use_sidepath then 10". Note that the current trekking profile uses a penalty of 4 for bicycle=no and no penalty for bicycle=sidepath

**Comment by Phyks** [comment]
> Without artifacts? It depends on what you call this situation

This is probably forbidden indeed and to be considered as an artifact, but likely an error (actually incomplete) errors in OpenStreetMap data. So, it somehow makes sense.

I would personnally consider it fine to have such artifacts for an optional (default would be off) variant of the trekking profile.

**Comment by polyscias** [comment]
I have called the profile I am using urban as traffic signals are something typically urban I think.

Yes, I think the code could be added to the trekking profile with default off but an urban profile makes also sense to me.

I am still wondering what to do with the bicycle=no|use_sidepath penalties, my first try to get bicycle=use_sidepath to use a penalty of 4 (#191) was not seen as valid, but for this it is essential to have a decent penalty. 

I checked the crossings mentioned in #79 and with a penalty of 4 but still a detour is taken over the main road, checked what is the minimum penalty for the two crossing and see it is 8/9, as the values is depend on the crossing and the penalty for traffic lights I think the 10 is still a conservative value.
